File: src/year2019/day1/day1.py
#!/usr/bin/env python3
# [Day 1: The Tyranny of the Rocket Equation](https://adventofcode.com/2019/day/1)

import atexit
import sys
import time
from pathlib import Path

filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text()
if "--elapsed" in sys.argv:
    sys.argv.remove("--elapsed")
    start_time_ns = time.time_ns()
    atexit.register(lambda: print(f"elapsed: {(time.time_ns() - start_time_ns) / 1_000_000}ms"))


lines = data.splitlines()

print(sum(int(mass) // 3 - 2 for mass in lines))

part2 = 0
for mass in lines:
    fuel = int(mass)
    while True:
        fuel = fuel // 3 - 2
        if fuel <= 0:
            break
        part2 += fuel
print(part2)

File: src/year2019/day1/day1.c
// [Day 1: The Tyranny of the Rocket Equation](https://adventofcode.com/2019/day/1)
// cc -o day1 -Wall -O2 day1.c

#include <stdio.h>

int main(int argc, char *argv[])
{
    FILE *f = fopen((argc >= 2) ? argv[1] : "input.txt", "r");

    int mass, part1 = 0, part2 = 0;
    while (fscanf(f, "%d", &mass) == 1)
    {
        part1 += mass / 3 - 2;

        int fuel = mass;
        while ((fuel = fuel / 3 - 2) > 0)
        {
            part2 += fuel;
        }
    }

    fclose(f);

    printf("%d\n", part1);
    printf("%d\n", part2);

    return 0;
}
